### PR TITLE
[MUGEN] MultimodalGPT: Make batch dim optional for attn masks, add num in/out tokens attributes

### PR DESCRIPTION
**Pull request by langong347**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #282
* #281
* #276
* __->__ #280

Summary:
- Allow users to pass the attention and head masks without creating the batch dim
- Add logic inside the GPT model stack to create batch dim if not supplied
- Add additional attributes to GPT model for input/output token states

Test Plan:
CI

Differential Revision: [D38932534](https://our.internmc.facebook.com/intern/diff/D38932534)

### CODE REVIEW

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38932534).

File: torchmultimodal/models/gpt.py [hunk 1]
```diff
@@ -475,6 +491,14 @@ def forward(
         return_attn_weights: bool = False,
         return_hidden_states: bool = False,
     ) -> TransformerDecoderOutput:
+        if attn_mask is not None and attn_mask.dim() == 2:
```
Comment (ebsmothers) [inline torchmultimodal/models/gpt.py:494]: Yeah I now better understand some of your comments around `get_extended_attention_mask`. Agree that it would be nice to find a way to support this case as well (since the end result is the same), but not necessary for this PR.

File: torchmultimodal/models/gpt.py [hunk 2]
```diff
@@ -475,6 +491,14 @@ def forward(
...
+            attn_mask = attn_mask[
+                None, None, :, :
+            ]  # (q_seq_len, k_seq_len) -> (b, h, q_seq_len, k_seq_len)
```
Comment (ebsmothers) [inline torchmultimodal/models/gpt.py:497]: nit: isn't it technically this?
```suggestion
            ]  # (q_seq_len, k_seq_len) -> (1, 1, q_seq_len, k_seq_len)
```

**Review by ebsmothers — APPROVED** [review]

File: torchmultimodal/models/gpt.py (hunk 2 above)
Comment (langong347) [reply torchmultimodal/models/gpt.py:497]: yeah I haven't been consistent with my notations. Let me update them.

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38932534).
